x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/correlation/CountCorrelationFunction.java

```diff
@@ -0,0 +1,162 @@
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class CountCorrelationFunction implements CorrelationFunction {
+
+    public static final ParseField NAME = new ParseField("count_correlation");
+    public static final ParseField INDICATOR = new ParseField("indicator");
+
+    private static final ConstructingObjectParser<CountCorrelationFunction, Void> PARSER = new ConstructingObjectParser<>(
+        "count_correlation_function",
+        false,
+        a -> new CountCorrelationFunction((CorrelativeValue)a[0])
+    );
+
+    static {
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> CorrelativeValue.fromXContent(p), INDICATOR);
+    }
+
+    private final CorrelativeValue indicator;
+
+    public CountCorrelationFunction(CorrelativeValue indicator) {
+        this.indicator = indicator;
+    }
+
+    public CountCorrelationFunction(StreamInput in) throws IOException {
+        this.indicator = new CorrelativeValue(in);
+    }
+
+    public static CountCorrelationFunction fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDICATOR.getPreferredName(), indicator);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        indicator.writeTo(out);
+    }
+
+    @Override
+    public String getName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public int hashCode() {
+        return NAME.getPreferredName().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        CountCorrelationFunction other = (CountCorrelationFunction) obj;
+        return Objects.equals(indicator, other.indicator);
+    }
+
+    @Override
+    public double execute(CorrelativeValue y) {
+        if (indicator.getExpectations().length != y.getExpectations().length) {
+            throw new AggregationExecutionException(
+                "value lengths do not match; indicator.expectations ["
+                    + indicator.getExpectations().length
+                    + "] and number of buckets ["
+                    + y.getExpectations().length
+                    + "]. Unable to calculate correlation"
+            );
+        }
+        final double xMean;
+        final double xVar;
+        if (indicator.getFractions() == null) {
+            double sum = CorrelationFunction.sum(indicator.getExpectations());
+            xMean = sum / indicator.getExpectations().length;
+            double var = 0;
```

Comment:
There is `MovingFunctions.stdDev` which calculates this as part of it. I wonder if it makes sense to add a `MovingFunctions.variance` method and reuse it from `MovingFunctions.stdDev`. If that feels a bit risky, I still think we might want to consider `NaN` handling here.